### PR TITLE
Add Aten::avg_pool3d forward/backward

### PR DESCRIPTION
**Pull request by chunhuanMeng**
Op:
- [x] Aten::avg_pool3d forward
- [x] Aten::avg_pool3d backward

Related issue:
https://github.com/intel/torch-xpu-ops/issues/771

case fallowing passed:
* nn/test_pooling_xpu.py::TestPoolingNNDeviceTypeXPU::test_pooling_bfloat16_xpu
* nn/test_pooling_xpu.py::TestPoolingNNDeviceTypeXPU::test_pool_large_size_xpu_bfloat16

### CODE REVIEW

**Comment by chunhuanMeng** [comment]
image

**Comment by chunhuanMeng** [comment]
image

File: src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp [hunk 1]
```diff
@@ -0,0 +1,979 @@
...
+#include <ATen/core/Tensor.h>
+#include <ATen/native/CanUse32BitIndexMath.h>
+#include <ATen/native/Pool.h>
+#include <ATen/native/xpu/sycl/Atomics.h>
+
+#include <ATen/native/xpu/sycl/AveragePool2dKernels.h>
+#include <comm/Runtime.h>
+#include <comm/SYCLContext.h>
+#include <comm/SYCLHelpers.h>
+#include "c10/core/DeviceType.h"
```
Comment (majing921201) [inline src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp:18]: Do we need this head file ?

File: src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp [hunk 2]
```diff
@@ -0,0 +1,979 @@
...
+#include <ATen/ceil_div.h>
+#include <ATen/core/Tensor.h>
+#include <ATen/native/CanUse32BitIndexMath.h>
+#include <ATen/native/Pool.h>
+#include <ATen/native/xpu/sycl/Atomics.h>
+
+#include <ATen/native/xpu/sycl/AveragePool2dKernels.h>
+#include <comm/Runtime.h>
+#include <comm/SYCLContext.h>
+#include <comm/SYCLHelpers.h>
```
Comment (majing921201) [inline src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp:17]: comm/SYCLContext.h includes others two head files

File: src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp [hunk 3]
```diff
@@ -0,0 +1,977 @@
...
+
+  if (input.ndimension() == 5) {
+    // Collapse batch and feature dimensions.
+    work_input = work_input.reshape({nbatch * nslices, itime, iheight, iwidth});
+    work_output =
+        work_output.reshape({nbatch * nslices, otime, oheight, owidth});
+  }
+  AT_DISPATCH_FLOATING_TYPES_AND2(
+      kHalf, kBFloat16, input.scalar_type(), "avg_pool3d_out_xpu", [&] {
+        using accscalar_t = acc_type_device<scalar_t, kXPU>;
```
Comment (majing921201) [inline src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp:255]: "avg_pool3d_xpu", it is a kernel impl, not related variants

File: src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp [hunk 4]
```diff
@@ -0,0 +1,977 @@
...
+        int64_t totalZ = otime * nslices * nbatch;
+        int64_t offsetZ = 0;
+        while (totalZ > 0) {
+          if (canUse32BitIndexMath(input)) {
+            avg_pool3d_out_frame<scalar_t, accscalar_t, int32_t>(
```
Comment (majing921201) [inline src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp:260]: CUDA avg_pool3d_out_frame means a CUDA kernel, not a launcher, here we can name is avg_pool3d_out_template or launch_avg_pool3d_kernel

File: src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp [hunk 5]
```diff
@@ -0,0 +1,977 @@
...
+    work_grad_output =
+        work_grad_output.reshape({nbatch * nslices, otime, oheight, owidth});
+  }
+  if (canUse32BitIndexMath(work_grad_input)) {
+    if (dT == 1 && dH == 1 && dW == 1 && padT == 0 && padH == 0 && padW == 0) {
+      AT_DISPATCH_FLOATING_TYPES_AND2(
+          kHalf,
+          kBFloat16,
+          input.scalar_type(),
+          "avg_pool3d_backward_out_frame_stride1",
```
Comment (majing921201) [inline src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp:795]: avg_pool3d_backward_stride1

File: src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp [hunk 6]
```diff
@@ -0,0 +1,977 @@
...
+            accscalar_t divide_factor;
+            if (divisor) {
+              divide_factor = static_cast<accscalar_t>(divisor);
+            } else {
+              divide_factor = static_cast<accscalar_t>(kT * kH * kW);
+            }
+
+            while (totalZ > 0) {
+              avg_pool3d_backward_out_frame_stride1<
+                  scalar_t,
```
Comment (majing921201) [inline src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp:810]: ditto

File: src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp [hunk 7]
```diff
@@ -0,0 +1,977 @@
...
+              totalZ -= 65535;
+              offsetZ += 65535;
+            }
+          });
+    } else {
+      AT_DISPATCH_FLOATING_TYPES_AND2(
+          kHalf,
+          kBFloat16,
+          input.scalar_type(),
+          "avg_pool3d_backward_out_frame",
```
Comment (majing921201) [inline src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp:830]: "avg_pool3d_backward_xpu"

File: src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp [hunk 8]
```diff
@@ -0,0 +1,977 @@
...
+          [&] {
+            using accscalar_t = acc_type_device<scalar_t, kXPU>;
+            int64_t totalZ = otime * nslices * nbatch;
+            int64_t offsetZ = 0;
+            while (totalZ > 0) {
+              if (kernelsOverlap) {
+                avg_pool3d_backward_out_frame_atomic<
```
Comment (majing921201) [inline src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp:837]: ditto

File: src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp [hunk 9]
```diff
@@ -0,0 +1,977 @@
...
+            }
+          });
+    }
+  } else {
+    if (dT == 1 && dH == 1 && dW == 1 && padT == 0 && padH == 0 && padW == 0) {
+      AT_DISPATCH_FLOATING_TYPES_AND2(
+          kHalf,
+          kBFloat16,
+          input.scalar_type(),
+          "avg_pool3d_backward_out_frame_stride1",
```
Comment (majing921201) [inline src/ATen/native/xpu/sycl/AveragePool3dKernels.cpp:886]: ditto

**Review by xytintel — APPROVED** [review]

**Pull request merged**